### ISSUE TITLE
Tasks are running twice - potential solution: use pytz.timezone.localize instead of tzinfo in datetime

### ISSUE THREAD

**Issue by julesbond007** [sibson/redbeat#160]
Hey there,

Thanks for the nice library. It's been pretty helpful. I have a question, like a lot of people say, the tasks run twice and I'm not sure if the bug is in the library or celery or redis, but I noticed how the schedule due datetime is calculated and you're passing tzinfo=self.schedule.tz, in my testing, I didn't find this reliable. Instead I think you may want to use:

`pytz.timezone('whatever_tz').localize(whatever_date_time)` 

Here is what I have in my testing:

With tzinfo (wrong):
```
>>> import datetime
>>> import pytz
>>> d = datetime.datetime(year=2020, month=4, day=2, hour=10, minute=0, tzinfo=pytz.timezone('America/New_York'))
>>> print(d)
2020-04-02 10:00:00-04:56
>>> y = d.astimezone(pytz.UTC)
>>> print(y)
2020-04-02 14:56:00+00:00
```

Using pytz.localize (correct):
```
>>> import datetime
>>> import pytz
>>> d = datetime.datetime(year=2020, month=4, day=2, hour=10, minute=0)
>>> d = pytz.timezone('America/New_York').localize(d)
>>> print(d)
2020-04-02 10:00:00-04:00
>>> z = d.astimezone(pytz.UTC)
>>> print(z)
2020-04-02 14:00:00+00:00
```

Celery uses astimezone to calculate now:

```
class Celery(...)
...
    def now(self):
        """Return the current time and date as a datetime."""
        now_in_utc = to_utc(datetime.utcnow())
        return now_in_utc.astimezone(self.timezone)
```

What I believe is happening is celery will have 1 time, and celery-redbeat will potentially give another time (not always) so, the task will run twice sometimes. If you schedule a task to run say at 12:00 in America/New_York timezone, it will be picked up also right around 11:56 a few minutes before. The time difference changes based on timezone offset.

See this Stack overflow as well:
https://stackoverflow.com/questions/6410971/python-datetime-object-show-wrong-timezone-offset

**Comment by sibson** [comment]
Redbeat isn't timezone aware and expects to run in UTC. At one point I thought that I'd documented that but couldn't find it when looking just now. It would be great if Redbeat was TZ aware but it's complicated enough that I'd need to spend more time than I have available to get it right.  There are a couple of open PRs that move things in the right direction but they have made incompatible tradeoffs.

**Comment by julesbond007** [comment]
I see! Thanks Marc!

**Comment by julesbond007** [comment]
Hey Marc, I'll close this issue for now.

**Comment by alxnik** [comment]
> Redbeat isn't timezone aware and expects to run in UTC. At one point I thought that I'd documented that but couldn't find it when looking just now. It would be great if Redbeat was TZ aware but it's complicated enough that I'd need to spend more time than I have available to get it right. There are a couple of open PRs that move things in the right direction but they have made incompatible tradeoffs.

@sibson I have the same issue although I work exclusively in UTC. Basically, the first invocation of each task runs 5 minutes early. All next invocations (where last_run is set) work ok.

Although #165 probably gives a final solution to the problem anyway, my understanding is that it will be released in v2.0.0 which introduces breaking changes (and is not released yet in pypi). I think it would be better to release a v1.0.1 which fixes this issue as per #161 (I personally tested it and it seems to work) and release #165 in v2.0.0 as already planned.

**Comment by sibson** [comment]
I wonder if https://github.com/sibson/redbeat/issues/210, was also contributing.
